### PR TITLE
[bc-breaking] reference option for conv produce a pattern instead of reference conv module

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62215
* __->__ #61942

Summary:
This PR changes is_reference=True for conv to produce a pattern consists of dequant - float conv - quant instead of reference conv module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810656](https://our.internmc.facebook.com/intern/diff/D29810656)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

File: torch/nn/intrinsic/qat/modules/conv_fused.py [hunk 1]
```diff
@@ -303,6 +303,32 @@ def forward(self, input):
     def from_float(cls, mod):
         return super(ConvBnReLU1d, cls).from_float(mod)
 
+    def to_float(self):
```
Comment (vkuzo) [inline torch/nn/intrinsic/qat/modules/conv_fused.py:306]: would it be possible to reuse code between 1d/2d/3d same as `from_float` is doing? Or is there some blocker

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

File: torch/nn/intrinsic/qat/modules/conv_fused.py (hunk 1 above)
Comment (jerryzh168) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:306]: probably not, let me try if I can combine them

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

File: torch/nn/intrinsic/qat/modules/conv_fused.py [hunk 2]
```diff
@@ -344,6 +388,24 @@ def __init__(self,
                            padding, dilation, False, _pair(0), groups, bias, padding_mode,
                            eps, momentum, freeze_bn, qconfig, dim=2)
 
+    def to_float(self):
```
Comment (vkuzo) [inline torch/nn/intrinsic/qat/modules/conv_fused.py:391]: is there a way for this to call `_ConvBnNd.to_float` to reuse more code?

Comment (jerryzh168) [reply torch/nn/intrinsic/qat/modules/conv_fused.py:391]: oh sorry, I have implemented this in ConvBnNd, forgot to delete the extra code

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

File: torch/quantization/fx/quantization_patterns.py [hunk 3]
```diff
@@ -526,25 +526,83 @@ def convert(self,
...
+                if activation_int8_quantized:
+                    dtype = activation_dtype(qconfig)
+                activation = load_arg(quantized=dtype)(self.conv_node.args[0])
+                args = load_arg(quantized=torch.float)(self.conv_node.args)
+                # Get the float conv and attach weight_post_process
+                # lowering pass can call weight_post_process.calculate_qparams
+                # to get scale and zero_point for weight
+                if isinstance(
+                        self.conv,
+                        (torch.nn.qat.Conv2d,
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:544]: nit, could this be a mapping specified on top of the file and used as a var here?  Just to reduce LOC in this function

File: torch/quantization/fx/quantization_patterns.py [hunk 4]
```diff
@@ -526,25 +526,83 @@ def convert(self,
...
+                         torch.nn.qat.Conv3d,
+                         torch.nn.intrinsic.qat.ConvBn2d,
+                         torch.nn.intrinsic.qat.ConvBnReLU2d,
+                         torch.nn.intrinsic.qat.ConvReLU2d,
+                         torch.nn.intrinsic.qat.ConvBn3d,
+                         torch.nn.intrinsic.qat.ConvBnReLU3d,
+                         torch.nn.intrinsic.qat.ConvReLU3d)):
+                    # case 1. converting qat conv module to
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:552]: thanks for the comments!

**Review by vkuzo — APPROVED** [review]

File: torch/quantization/fx/quantization_patterns.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:544]: sure

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810656).

**Pull request closed without merging**